### PR TITLE
add e2e tests

### LINKED ISSUE #5750: E2E test scenario 1 - Starting from an empty directory
/kind user-story

User Story
As a quality engineer, I want to create e2e tests that involve using all the features to simulate the developer workflow.

Scenario 1 - Start `odo init` in an empty directory:
Connect to cluster
Run `odo init` in interactive mode select javascript/nodejs, then take all default values
Expect output to include "Your new component 'my-nodejs-app' is ready in the current directory."

run `odo dev `
Expect to get the project deployes to the cluster in developer mode, and the application should get running in the cluster, finally it should enter watch mode waiting for changes to files to redeploy automatically.

Make changes to the server.js file and save them.
Verify to get changes pushed to the server, verify that the changes got deployed.

Make additional changes and save again. 
Verify that the additional changes got deployed to the cluster.

Exit dev mode

Run `odo deploy` 
Verify that the component was successfully deployed by running `odo list`

Enter dev mode again

Make additional changes. 

Exit dev mode

Run `odo deploy`
Verify that the additional changes were successfully redeployed.

Undeploy the application form the cluster by running `odo delete component`.
Expect the command to ask for confirmation to delete the component.
Verify that the command confirms the deletion.
Verify that the component wad deleted from the cluster.

### LINKED ISSUE #5752: E2E test scenario 2 - Starting from an directory with sources
/kind user-story

User Story
As a quality engineer, I want to create e2e tests that involve using all the features to simulate the developer workflow.

Scenario 2 - Start `odo init` from a directory with sources:
Connect to cluster
Run `odo init` in interactive mode select javascript/nodejs, then take all default values
Expect output to include "Your new component 'my-nodejs-app' is ready in the current directory."

run `odo dev `
Expect to get the project deployes to the cluster in developer mode, and the application should get running in the cluster, finally it should enter watch mode waiting for changes to files to redeploy automatically.

Make changes to the server.js file and save them.
Verify to get changes pushed to the server, verify that the changes got deployed.

Make additional changes and save again. 
Verify that the additional changes got deployed to the cluster.

Exit dev mode

Run `odo deploy` 
Verify that the component was successfully deployed by running `odo list`

Enter dev mode again

Make additional changes. 

Exit dev mode

Run `odo deploy`
Verify that the additional changes were successfully deployed.

Undeploy the application form the cluster by running `odo delete component`.
Expect the command to ask for confirmation to delete the component.
Verify that the command confirms the deletion.
Verify that the component wad deleted from the cluster.

### PR DESCRIPTION
**Pull request by anandrkskd**
Signed-off-by: anandrkskd <anandrkskd@gmail.com>

/kind tests

**What does this PR do / why we need it:**
This PR adds e2e tests for odo.

**Which issue(s) this PR fixes:**

Fixes #5752
Fixes #5750 

**PR acceptance criteria:**

- [ ] Unit test 

- [x] Integration test 

- [ ] Documentation 

**How to test changes / Special notes to the reviewer:**

### CODE REVIEW

File: tests/e2escenarios/e2e-test.go [hunk 1]
```diff
@@ -0,0 +1,258 @@
...
+
+			// "execute odo dev and add changes to application"
+			var devSession helper.DevSession
+			var outContents []byte
+			var ports map[string]string
+
+			devSession, outContents, _, ports, err = helper.StartDevMode()
+			helper.ReplaceString(filepath.Join(commonVar.Context, "server.js"), "from Node.js", "from updated Node.js")
+			Expect(err).ToNot(HaveOccurred())
+			helper.MatchAllInOutput(string(outContents), []string{"Watching for changes in the current directory"})
```
Comment (feloy) [inline tests/e2escenarios/e2e-test.go:84]: Here, you are checking for the first "Watching for changes in the current directory", not the one after the sync of server.js happens.

You should use devSession.Waitsync() instead

File: tests/e2escenarios/e2e-test.go [hunk 2]
```diff
@@ -0,0 +1,258 @@
...
+
+			// "running odo list"
+			stdout := helper.Cmd("odo", "list").ShouldPass().Out()
+			helper.MatchAllInOutput(stdout, []string{componentName, "nodejs", "Dev"})
+
+			// "exit dev mode and run odo deploy"
+			devSession.Stop()
+			devSession.WaitEnd()
+
+			// "run odo deploy"
```
Comment (feloy) [inline tests/e2escenarios/e2e-test.go:104]: You could check at this point that there are no more resources in the namespace

File: tests/e2escenarios/e2e-test.go [hunk 3]
```diff
@@ -0,0 +1,258 @@
...
+			helper.CopyExampleDevFile(filepath.Join("source", "devfiles", "nodejs", "devfile-deploy.yaml"), path.Join(commonVar.Context, "devfile.yaml"))
+			helper.ReplaceString(filepath.Join(commonVar.Context, "devfile.yaml"), "nodejs-prj1-api-abhz", componentName)
+
+			//helper.ReplaceString(filepath.Join(commonVar.Context, "devfile.yaml"), "quay.io/unknown-account/myimage", "quay.io/repository/libpod/alpine")
+			stdout = helper.Cmd("odo", "deploy").AddEnv("PODMAN_CMD=echo").ShouldPass().Out()
```
Comment (feloy) [inline tests/e2escenarios/e2e-test.go:109]: It could be interesting to build the image for the e2e tests

**Review by feloy — CHANGES_REQUESTED** [review]

File: tests/e2escenarios/e2e-test.go (hunk 2 above)
Comment (anandrkskd) [reply tests/e2escenarios/e2e-test.go:103]: By "resources", do you mean deployment ? 
or how do we do it?

File: tests/e2escenarios/e2e-test.go (hunk 3 above)
Comment (feloy) [reply tests/e2escenarios/e2e-test.go:108]: Not for this PR, but we should discuss this during a cabal

File: Makefile [hunk 4]
```diff
@@ -102,7 +102,7 @@ vet:
 
 .PHONY: sec
 sec:
-	go run $(COMMON_GOFLAGS) github.com/securego/gosec/v2/cmd/gosec -severity medium -confidence medium -exclude G304,G204 -quiet  ./...
+	go run $(COMMON_GOFLAGS) github.com/securego/gosec/v2/cmd/gosec -severity medium -confidence medium -exclude G304,G204,G107 -quiet  ./...
```
Comment (feloy) [inline Makefile:105]: I would prefer we disable the G107 rule for tests only, by running different commands for odo and for tests:

```
go run $(COMMON_GOFLAGS) github.com/securego/gosec/v2/cmd/gosec -severity medium -confidence medium -exclude G304,G204,G107 -quiet  ./tests/...
go run $(COMMON_GOFLAGS) github.com/securego/gosec/v2/cmd/gosec -severity medium -confidence medium -exclude G304,G204 -quiet  ./cmd/... ./pkg/...
```

**Review by feloy — CHANGES_REQUESTED** [review]

File: Makefile (hunk 4 above)
Comment (anandrkskd) [reply Makefile:105]: Understood!

File: tests/e2escenarios/e2e-test.go [hunk 5]
```diff
@@ -0,0 +1,275 @@
...
+			pvcs := commonVar.CliRunner.GetAllPVCNames(commonVar.Project)
+			Expect(pvcs).To(BeEmpty())
+			pods := commonVar.CliRunner.GetAllPodNames(commonVar.Project)
+			Expect(pods).To(BeEmpty())
+
+			// "run odo deploy"
+			helper.CopyExampleDevFile(filepath.Join("source", "devfiles", "nodejs", "devfile-deploy.yaml"), path.Join(commonVar.Context, "devfile.yaml"))
+			helper.ReplaceString(filepath.Join(commonVar.Context, "devfile.yaml"), "nodejs-prj1-api-abhz", componentName)
+
+			//helper.ReplaceString(filepath.Join(commonVar.Context, "devfile.yaml"), "quay.io/unknown-account/myimage", "quay.io/repository/libpod/alpine")
```
Comment (feloy) [inline tests/e2escenarios/e2e-test.go:114]: Please remove this commented code

File: tests/e2escenarios/e2e-test.go [hunk 6]
```diff
@@ -0,0 +1,274 @@
...
+				helper.SendLine(ctx, "nodejs-starter\n")
+
+				helper.ExpectString(ctx, "Enter component name")
+				helper.SendLine(ctx, componentName)
+
+				helper.ExpectString(ctx, "Your new component '"+componentName+"' is ready in the current directory")
+
+			})
+			Expect(err).To(BeNil())
+			Expect(output).To(ContainSubstring("Your new component '" + componentName + "' is ready in the current directory"))
```
Comment (valaparthvi) [inline tests/e2escenarios/e2e-test.go:71]: Why this check again?

File: tests/e2escenarios/e2e-test.go [hunk 7]
```diff
@@ -0,0 +1,274 @@
...
+
+			devSession, _, _, ports, err = helper.StartDevMode()
+			helper.ReplaceString(filepath.Join(commonVar.Context, "server.js"), "from Node.js", "from updated Node.js")
+			Expect(err).ToNot(HaveOccurred())
+			_, _, err = devSession.WaitSync()
+			Expect(err).ToNot(HaveOccurred())
+			// "should update the changes"
+			checkIfDevEnvIsUp(ports["3000"], "Hello from updated Node.js Starter Application!")
+
+			// "changes are made made to the applications"
```
Comment (valaparthvi) [inline tests/e2escenarios/e2e-test.go:87]:
```suggestion
			// "changes are made to the applications"
```

File: tests/e2escenarios/e2e-test.go [hunk 8]
```diff
@@ -0,0 +1,274 @@
...
+		componentName := helper.RandString(6)
+		var _ = BeforeEach(func() {
+			helper.Chdir(commonVar.Context)
+			helper.CopyExample(filepath.Join("source", "devfiles", "nodejs", "project"), commonVar.Context)
+		})
+		It("should verify developer workflow from non-empty Directory", func() {
+			deploymentName := "my-component"
+			serviceName := "my-cs"
+			getDeployArgs := []string{"get", "deployment", "-n", commonVar.Project}
+			getSVCArgs := []string{"get", "svc", "-n", commonVar.Project}
```
Comment (valaparthvi) [inline tests/e2escenarios/e2e-test.go:166]: You can perhaps move these to the main `Context` since it's common for both the sub-contexts.

Comment (anandrkskd) [reply tests/e2escenarios/e2e-test.go:166]: Previously, Philippe recommended that we should define the these value close to tests where it's getting used. 
cc @feloy

File: tests/e2escenarios/e2e-test.go (hunk 6 above)
Comment (anandrkskd) [reply tests/e2escenarios/e2e-test.go:71]: Removed, That check was redundant.

**Comment by valaparthvi** [comment]
/override prow/v4.10-integration-e2e

IBM tests pass.

File: tests/e2escenarios/e2e-test.go (hunk 3 above)
Comment (anandrkskd) [reply tests/e2escenarios/e2e-test.go:113]: I will create an issue for this one separately to track.

**Pull request merged**